### PR TITLE
V2.1.0

### PR DESCRIPTION
**Pull request by sasbury**
[Added] Added named dispatchers to subscriptions to allow application control of callback threads
[#100] Split internal callback dispatchers to prevent application code blocking heartbeats

### CODE REVIEW

File: src/main/java/io/nats/streaming/StreamingConnectionImpl.java [hunk 1]
```diff
@@ -119,22 +123,32 @@ StreamingConnectionImpl connect() throws IOException, InterruptedException {
             this.hbSubject = this.newInbox();
             this.ackSubject = String.format("%s.%s", NatsStreaming.DEFAULT_ACK_PREFIX, this.nuid.next());
 
-            this.dispatcher = nc.createDispatcher(msg -> {
+            this.ackDispatcher = nc.createDispatcher(msg -> {
                 String subject = msg.getSubject();
+                if (this.ackSubject.equals(subject)) {
```
Comment (kozlovic) [inline src/main/java/io/nats/streaming/StreamingConnectionImpl.java]: If you create a dedicated dispatcher, why would you check the subject?

File: src/main/java/io/nats/streaming/StreamingConnectionImpl.java [hunk 2]
```diff
@@ -119,22 +123,32 @@ StreamingConnectionImpl connect() throws IOException, InterruptedException {
...
+                    this.processAck(msg);
+                }
+            });
 
+            this.heartbeatDispatcher = nc.createDispatcher(msg -> {
+                String subject = msg.getSubject();
```
Comment (kozlovic) [inline src/main/java/io/nats/streaming/StreamingConnectionImpl.java]: Same as above

File: src/main/java/io/nats/streaming/StreamingConnectionImpl.java [hunk 3]
```diff
@@ -231,11 +245,24 @@ public void close() throws IOException, InterruptedException {
                 }
                 ackTimer.cancel();
 
+                if (this.messageDispatcher != null && this.messageDispatcher.isActive()) {
+                    nc.closeDispatcher(this.messageDispatcher);
```
Comment (kozlovic) [inline src/main/java/io/nats/streaming/StreamingConnectionImpl.java]: for the other dispatchers (ack and hb), you unsubscribe, but not here. Is there a reason?

File: src/test/java/io/nats/streaming/SubscribeTests.java [hunk 4]
```diff
@@ -870,4 +926,47 @@ public void testNoDuplicatesOnSubscriberStart() throws Exception {
...
+                    }
+                };
+
+                // Should receive all messages, and heartbeat should be ok despite sleeps in callback.
+                try (Subscription sub = sc.subscribe("foo", mcb,
+                        new SubscriptionOptions.Builder().deliverAllAvailable().build())) {
+
+                    assertTrue("Did not receive our messages", latch.await(3 * count * sleep, TimeUnit.SECONDS));
+                    assertEquals("Got wrong number of msgs", count, received.get());
+                }
```
Comment (kozlovic) [inline src/test/java/io/nats/streaming/SubscribeTests.java]: You are not really checking that the client has not been closed by the server here. So what I would do is initially send a single message, sleep in the callback for more than the server would need to close the client if the HB was blocked by callback.
After that time, send a new message and this time you don't need to sleep in callback. Test is OK if callback received 2 messages.

File: src/main/java/io/nats/streaming/StreamingConnectionImpl.java [hunk 5]
```diff
@@ -119,22 +123,32 @@ StreamingConnectionImpl connect() throws IOException, InterruptedException {
...
-            this.dispatcher.subscribe(this.hbSubject);
-            this.dispatcher.subscribe(this.ackSubject);
+            this.messageDispatcher = nc.createDispatcher(msg -> {
+                this.processMsg(msg);
+            });
+
+            this.heartbeatDispatcher.subscribe(this.hbSubject);
+            this.ackDispatcher.subscribe(this.ackSubject);
+
+            this.heartbeatDispatcher.setPendingLimits(-1, -1);
```
Comment (ColinSullivan1) [inline src/main/java/io/nats/streaming/StreamingConnectionImpl.java]: Not part of this PR, but maybe someday this would need to reference the NATS options.

Comment (sasbury) [reply src/main/java/io/nats/streaming/StreamingConnectionImpl.java]: i am not sure what you mean?

File: src/main/java/io/nats/streaming/SubscriptionOptions.java [hunk 6]
```diff
@@ -191,7 +206,8 @@ public int hashCode() {
         result = 31 * result + (ackWait != null ? ackWait.hashCode() : 0);
         result = 31 * result + (startAt != null ? startAt.hashCode() : 0);
         result = 31 * result + (int) (startSequence ^ (startSequence >>> 32));
-        result = 31 * result + (startTime != null ? startTime.hashCode() : 0);
+        result = 31 * result + (startTime != null ? startTime.getNano() : 0);
```
Comment (ColinSullivan1) [inline src/main/java/io/nats/streaming/SubscriptionOptions.java]: Just curious - why override hashCode?  For performance maybe save off the value after first build so it doesn't need to be rebuilt every time it's called.

Comment (sasbury) [reply src/main/java/io/nats/streaming/SubscriptionOptions.java]: this is one of those places where i am trying to minimize the changes for this release. It was overridden before, so I just added the minimum to remain correct. I suspect there is little if any value in this as the options aren't used as keys anywhere.

File: src/main/java/io/nats/streaming/StreamingConnectionImpl.java (hunk 5 above)
Comment (ColinSullivan1) [reply src/main/java/io/nats/streaming/StreamingConnectionImpl.java]: Should pending limits always be infinite?

Comment (sasbury) [reply src/main/java/io/nats/streaming/StreamingConnectionImpl.java]: oh, agreed, we may want to surface these in the future. I think for ack and hb we may want to keep hidden. But for message queues there is definitely value in making visible - with the caveat that these will interact with other streaming settings. So, may not be something we ever get to if we can get jetstream going.
